### PR TITLE
fix(rust): handle zero-size objects in download_multipart_to_bytes

### PR DESCRIPTION
**Pull request by joshuarobinson**
## Description

`file_size - 1` / `length - 1` on a `u64` underflowed for empty objects
and `Range(offset, size=0)`: a debug build panicked (surfacing in Python
as a `BaseException`-derived `PanicException`) and a release build
wrapped to `u64::MAX` and issued `get_range(0..0)`, which object_store
rejects. Return an empty buffer before computing the inclusive end
offset, matching `download_multipart_to_file`, which already tolerates
zero-size objects.

Release note: `RustClient.download_multipart_to_bytes` now returns `b""` for zero-byte objects and empty ranges instead of failing with an integer underflow / invalid range error.

_Found during a review of the commit history; no tracked issue._

## Checklist

- Development PR
  - `.release_notes/.unreleased.md`
    - [x] Notable changes to the client (i.e. not related to tooling, CI/CD, etc.) from this PR have been added.

## Testing

Compile-checked with `cargo check`; behaviour is exercised by the existing Rust client tests (simulator-backed).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed multipart downloads for zero-byte objects and empty ranges, which now return an empty byte string instead of producing errors.
  * Prevented invalid or overflowing range calculations from causing unexpected download failures.
  * Preserved existing behavior for full-object downloads and ranged requests, including the appropriate request handling for each download type.
  * Downloads now report a storage error before attempting a request when range arithmetic cannot be safely completed.

### CODE REVIEW

File: multi-storage-client/rust/src/lib.rs [hunk 1]
```diff
@@ -884,19 +884,22 @@ impl RustClient {
...
-                (0, file_size - 1, file_size)
+                (0, result.size)
             };
 
+            // Zero-byte objects (or empty ranges) have nothing to fetch; return before the
+            // inclusive end offset below would underflow.
+            if total_size == 0 {
+                return Ok(PyBytes::new(bytes::Bytes::new()));
+            }
+            let end_offset = start_offset + total_size - 1;
```
[the remark this answers is not in the document]
Comment (joshuarobinson) [reply multi-storage-client/rust/src/lib.rs]: Agreed. The explicit-range branch now uses `checked_add` and rejects a range whose exclusive end would not fit in a `u64`, returning `StorageError::ObjectStoreError("Invalid byte range: ...")` before any request is made.

**Review by dreamtalen — APPROVED** [review]
LGTM, it would be good to add a regression test that uploads a zero-byte object with `upload_multipart_from_bytes`, then downloads with `download_multipart_to_bytes`.
